OemPkg/OemConfigPolicyCreatorPei/OemConfigPolicyCreatorPei.c

```diff
@@ -0,0 +1,256 @@
+      DEBUG ((
+        DEBUG_ERROR,
+        "%a variable services failed to find %a with status (%r)\n",
+        __FUNCTION__,
+        gKnobData[i].Name,
+        Status
+        ));
+      ASSERT (FALSE);
+      FreePool (*ConfPolicy);
+      *ConfPolicySize = 0;
+      return Status;
+    }
+
+    DUMP_HEX(DEBUG_ERROR, 0, gKnobData[i].CacheValueAddress, sizeof(BOOLEAN), "OSDDEBUG: ");
+    ConfListPtr = ((UINT8 *)*ConfPolicy) + Offset;
+
+    VarListSize = NeededSize - Offset;
+    Status = ConvertVariableEntryToVariableList (&VarListEntry, ConfListPtr, &VarListSize);
+
+    if (EFI_ERROR (Status)) {
+      DEBUG((DEBUG_ERROR, "%a failed to convert variable entry to var list! - %r\n", __FUNCTION__, Status));
+      ASSERT (FALSE);
+      FreePool (*ConfPolicy);
+      return EFI_ABORTED;
+    }
+
+    Offset += VarListSize;
+  }
+
+  DUMP_HEX(DEBUG_ERROR, 0, *ConfPolicy, Offset, "OSDDEBUG: ");
+
+  if (Offset != NeededSize) {
+    // oops we messed up the math, may have corrupted memory...
+    DEBUG((DEBUG_ERROR, "%a expected ConfPolicy size %x does not match actual size %x!\n", __FUNCTION__, NeededSize, Offset));
+    ASSERT (Offset == NeededSize);
+    FreePool (*ConfPolicy);
+    return EFI_ABORTED;
+  }
+
+  DEBUG ((DEBUG_ERROR, "%a ConfPolicy: %p *ConfPolicy: %p ConfVarListPtr: %p &ConfListPtr: %p\n", __FUNCTION__, ConfPolicy, *ConfPolicy, ConfListPtr, &ConfListPtr));
+
+  return EFI_SUCCESS;
+}
+
+/**
+  Module entry point that will check configuration data and publish them to policy database.
+
+  @param FileHandle                     The image handle.
+  @param PeiServices                    The PEI services table.
+
+  @retval Status                        From internal routine or boot object, should not fail
+**/
+EFI_STATUS
+EFIAPI
+OemConfigPolicyCreatorPeiEntry (
+  IN EFI_PEI_FILE_HANDLE     FileHandle,
+  IN CONST EFI_PEI_SERVICES  **PeiServices
+  )
+{
+  EFI_STATUS  Status;
+  POLICY_PPI  *PolPpi        = NULL;
+  VOID        *ConfPolicy    = NULL;
+  UINT16       ConfPolicySize = 0;
+
+  DEBUG ((DEBUG_INFO, "%a - Entry.\n", __FUNCTION__));
+
+  // First locate policy ppi.
+  Status = PeiServicesLocatePpi (&gPeiPolicyPpiGuid, 0, NULL, (VOID *)&PolPpi);
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "%a Failed to locate Policy PPI - %r\n", __FUNCTION__, Status));
+    ASSERT (FALSE);
+    return Status;
+  }
+
+  // Oem can choose to do any Oem specific things to config here such as enforcing static only config or
+  // selecting a configuration profile based on some criteria
+
+  Status = CreateConfPolicy (&ConfPolicy, &ConfPolicySize);
+
+  DEBUG ((DEBUG_ERROR, "%a ConfPolicy: %p Size: %d\n", __FUNCTION__, ConfPolicy, ConfPolicySize));
+
+  if (EFI_ERROR (Status) || (ConfPolicy == NULL) || (ConfPolicySize == 0)) {
+    DEBUG ((DEBUG_ERROR, "%a CreateConfPolicy failed! Status (%r)\n", __FUNCTION__, Status));
+    ASSERT (FALSE);
+    return Status;

```

Comment:
Returning an error code from an Image Entry point can be problematic because the CoreStartImage can unload the the driver on an error code. 

https://github.com/microsoft/mu_basecore/blob/release/202208/MdeModulePkg/Core/Dxe/Image/Image.c#L1836

Reply:
Thanks for the info. I guess in this case we can return success and later consumers of the config policy can still come up and build their own error cases into not finding the config policy.